### ISSUE TITLE
Permission Issue: Cannot record video in Android Spectator App

### ISSUE THREAD

**Issue by BastiaanGrisel** [microsoft/MixedReality-SpectatorView#142]
I am trying to make a video of the BUILD2019 sample on my Pixel 3a Android phone. Both the Hololens and the Android app work, but pressing the record button on the Android phone does not start a recording.

### Steps
1. I followed [this readme](https://github.com/microsoft/MixedReality-SpectatorView/blob/7f9a0af04fdfd53bb8829d4ec23c20525f658c80/samples/Build2019Demo.Unity/README.md) and built the BUILD2019 sample on my Hololens 1. This worked fine.
4. For the Android app, simply running it using "Build and Run" showed the hologram that the Hololens user was seeing as well (so far so good). 
5. I brought up the menu on the Android app and pressed the red record button. A countdown starts but stops/freezes at 0. The screen stayed like this and tapping the stop button did not do anything:

<img src="https://user-images.githubusercontent.com/1348869/61804926-1c21fd00-ae35-11e9-901d-704d5c0d768b.png" width="50%">

5. I also tried exporting the project to Android Studio (instead of using "Build and Run") following [this readme](https://github.com/microsoft/MixedReality-SpectatorView/blob/7f9a0af04fdfd53bb8829d4ec23c20525f658c80/doc/SpectatorView.Setup.md#android) and had the same result. 

6. After tapping the stop button, no movie file was anywhere to be found on the Android phone when inspecting its contents through Windows Explorer. Looking at #122, the movie is saved to the SD card, while my Pixel 3a does not have an SD card slot.

### Debugging Efforts
1. When pressing the Record Button in the app, the LogCat logs say **Failed to initialize AndroidRecordingService: No errors**

![FailedToInitialiseAndroidRecordingDevice](https://user-images.githubusercontent.com/1348869/61805103-791db300-ae35-11e9-91a0-f3e18fa1523a.PNG)

2. Line 38 in `AndroidRecordingService.cs` logs this message. It does so because `Initialize` in `screenRecorderActivity` (which is the C# wrapper of `ScreenRecorderActivity.java`) returns **false** in line 35: 

![FailedToInitialiseARC](https://user-images.githubusercontent.com/1348869/61805218-b4b87d00-ae35-11e9-9e99-49d1d79fbc0b.PNG)

> BTW: the **No Errors** in **Failed to initialize AndroidRecordingService: No Errors** is user defined and does not mean there are no errors. It is returned by `GetLastErrorMessage()` in `AndroidRecordingService.cs` which calls a method in `ScreenRecorderActivity.java` that returns the `lastErrorMessage` (line 45) that defaults to `NO_ERROR_MESSAGE` (line 40) which is **No errors**:
>
> ![NoErrors](https://user-images.githubusercontent.com/1348869/61808235-4d053080-ae3b-11e9-8e89-4cae1007a13c.PNG)

3. Inspecting the `Initialize` method of `ScreenRecorderActivity.java` in Android studio, it returns **false** because the `state` is **ScreenRecordingActivity.NO_PERMISSIONS**, as shown by this Android Studio debug breakpoint (`ScreenRecorderActivity.java`, line 107): 

![NoPermissionsToRecord](https://user-images.githubusercontent.com/1348869/61805234-be41e500-ae35-11e9-93d7-4b1251879d80.PNG)

4. In order to set permissions, `requestPermissions` is called in the `onCreate` method of `ScreenRecorderActivity.java` as follows: 

![PermissionList](https://user-images.githubusercontent.com/1348869/61805245-c4d05c80-ae35-11e9-84ff-c7f73c98425c.PNG)

5.  `requestPermissions` triggers a callback `onRequestPermissionsResult` which sets the `state`. However, when `onRequestPermissionsResult` is called, the argument `grantResults` is **[0, -1]** [which means](https://developer.android.com/reference/android/content/pm/PackageManager.html#PERMISSION_DENIED) **[PERMISSION_GRANTED, PERMISSION_DENIED]**. This means that **Manifest.permission.WRITE_EXTERNAL_STORAGE** is **PERMISSION_DENIED**:

![onRequestPermissions](https://user-images.githubusercontent.com/1348869/61810757-3dd4b180-ae40-11e9-9cb7-55b19cff210e.PNG)

6. I tried setting the permissions in my Android settings but without results. I am using a Google Pixel 3a running Android 9.0:

<img src="https://user-images.githubusercontent.com/1348869/61817895-e63e4200-ae4f-11e9-92dd-0e520d5206d2.png" width="50%">

**Comment by chrisfromwork** [comment]
It looks like our samples documentation is stale for android. I think it was added before we had recording support. You do need to export the application to android studio in order to specify the ScreenRecorderActivity as the default activity. Otherwise the requests for android capabilities needed for screen recording won't get handled correctly.

**Comment by BastiaanGrisel** [comment]
Upon further digging, I found these permissions in the AnroidManifest:

![PermissionsAndroidManifest](https://user-images.githubusercontent.com/1348869/61944601-5d81ec00-af9e-11e9-93e9-3452f7a942ac.PNG)

When I remove the `android:maxSdkVersion="18"` from these two lines and build to Android, it does ask me for the permission to read/write files and indeed records a video. I can see it in the root folder when I connect my Pixel 3a to my PC. So this all seems to work now!

> It looks like our samples documentation is stale for android.

Yes, it might good to reference the [**Building & Deploying** section of **Spectating with an Android or iOS Device Setup**](https://github.com/microsoft/MixedReality-SpectatorView/blob/7f9a0af04fdfd53bb8829d4ec23c20525f658c80/doc/SpectatorView.Setup.md#building--deploying) under [**Build & Deploy** section of the **//BUILD 2019 Demo**](https://github.com/microsoft/MixedReality-SpectatorView/blob/7f9a0af04fdfd53bb8829d4ec23c20525f658c80/samples/Build2019Demo.Unity/README.md#build--deploy).

BTW: feel free to close this and my previous issue if you want. I don't know if you want to leave them open and connect them to any update-documentation TODOs or just close them for now and keep them for future reference.

**Comment by chrisfromwork** [comment]
First off, thank you for all of your help with troubleshooting these issues. We greatly appreciate seeing this feedback and will update documentation/code to reflect your suggestions.

For now, I think we can keep this issue open. It looks like we should remove the maxSdkVersion from the manifests we have declared. It also seems like we can update documentation to improve these experiences. We may rename the issue for easier tracking of whats still needed.

**Comment by chrisfromwork** [comment]
@BastiaanGrisel after taking a look at the maxSdkVersion arguments in the user-permission stubs, it doesn't appear that these maxSdkVersions are declared in our provided AndroidManifest.xml. I've added documentation to our troubleshooting section around removing these tags to see if it allows screen recording to work. I've also updated the Build 2019 sample to call out exporting the android project to android studio. Thanks again for this feedback!
